src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,2181 @@
+                      strHex(alice.pk());
+                jv[sfBatchSigners.jsonName][signer.index]
+                  [sfBatchSigner.jsonName][sfTxnSignature.jsonName] =
+                      strHex(Slice{sig.data(), sig.size()});
+            }
+
+            jv = addBatchSignatures(jv, signers);
+
+            env(jv, ter(temBAD_SIGNATURE));
+            env.close();
+        }
+
+        // temMALFORMED: Batch: duplicate TxID found.
+        {
+            auto const batchFee = calcBatchFee(env, 1, 2);
+            Json::Value jv =
+                batch::batch(alice, env.seq(alice), batchFee, tfAllOrNothing);
+
+            // Tx 1
+            Json::Value tx1 = pay(alice, bob, XRP(10));
+            jv = addBatchTx(jv, tx1, env.seq(alice) + 1);
+            auto txn1 = jv[jss::RawTransactions][0u][jss::RawTransaction];
+            STParsedJSONObject parsed1(std::string(jss::tx_json), txn1);
+            STTx const stx1 = STTx{std::move(parsed1.object.value())};
+
+            // Tx 2
+            Json::Value const tx2 = pay(bob, alice, XRP(5));
+            jv = addBatchTx(jv, tx2, env.seq(bob));
+
+            // Add a duplicate hash
+            jv[sfTransactionIDs.jsonName].append(
+                to_string(stx1.getTransactionID()));
+            jv[sfTransactionIDs.jsonName].append(
+                to_string(stx1.getTransactionID()));
+
+            env(jv, batch::sig(bob), ter(temMALFORMED));
+            env.close();
+        }
+
+        // temMALFORMED: Batch: order of inner transactions does not match
+        // TxIDs.
+        {
+            auto const batchFee = calcBatchFee(env, 1, 2);
+            Json::Value jv =
+                batch::batch(alice, env.seq(alice), batchFee, tfAllOrNothing);
+
+            // Tx 1
+            Json::Value tx1 = pay(alice, bob, XRP(10));
+            jv = addBatchTx(jv, tx1, env.seq(alice) + 1);
+            auto txn1 = jv[jss::RawTransactions][0u][jss::RawTransaction];
+            STParsedJSONObject parsed1(std::string(jss::tx_json), txn1);
+            STTx const stx1 = STTx{std::move(parsed1.object.value())};
+
+            // Tx 2
+            Json::Value const tx2 = pay(bob, alice, XRP(5));
+            jv = addBatchTx(jv, tx2, env.seq(bob));
+            auto txn2 = jv[jss::RawTransactions][1u][jss::RawTransaction];
+            STParsedJSONObject parsed2(std::string(jss::tx_json), txn2);
+            STTx const stx2 = STTx{std::move(parsed2.object.value())};
+
+            // Add the hashes out of order
+            jv[sfTransactionIDs.jsonName].append(
+                to_string(stx2.getTransactionID()));
+            jv[sfTransactionIDs.jsonName].append(
+                to_string(stx1.getTransactionID()));
+
+            env(jv, batch::sig(bob), ter(temMALFORMED));
+            env.close();
+        }
+
+        // temINVALID_BATCH: Batch: TransactionType missing in array entry.
+        {
+            auto const txBlob =
```

Comment:
This test is hard to audit. Can it be written to use the existing JSON structures, but instead of using `batch::add`, the JSON is constructed by hand?

Reply:
I was having difficulty with the Env and creating inner transactions without a TransactionType. I will try it again after some other fixes.

Reply:
fixed